driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -73,7 +75,17 @@ private void closeSessionIfCountIsZero() {
 
     @Override
     public ConnectionSource getReadConnectionSource() {
-        return new SessionBindingConnectionSource(wrapped.getReadConnectionSource());
+        if (session.hasActiveTransaction()) {
```

Comment:
This implements the read preference "pinning" behavior required by the spec: thou shalt not change read preference within a transaction